server/app/models/ApplicationEvent.java

```diff
@@ -0,0 +1,81 @@
+package models;
+
+import io.ebean.annotation.DbJson;
+import io.ebean.annotation.WhenCreated;
+import java.time.Instant;
+import javax.persistence.Entity;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+import play.data.validation.Constraints;
+import services.application.ApplicationEventDetails;
+
+@Entity
+@Table(name = "application_events")
+public class ApplicationEvent extends BaseModel {
+
+  // The Application the event is on.
+  @ManyToOne private Application application;
+  // The {@code ApplicationEventDetails.Type} of the event.
+  @Constraints.Required private ApplicationEventDetails.Type eventType;
+  // The Account that triggered the event.
+  @ManyToOne
+  @JoinColumn(name = "actor_id")
+  private Account actor;
+  // Details of the event specific to the eventType.
+  @Constraints.Required @DbJson private ApplicationEventDetails details;
+  @WhenCreated private Instant createTime;
+
+  public ApplicationEvent() {}
+
+  public ApplicationEvent(
+      Application application,
+      Account actor,
+      ApplicationEventDetails.Type eventType,
+      ApplicationEventDetails details) {
+    this.application = application;
```

Comment:
We don't do this in our other models, but I think we could definitely benefit from `checkNotNull()` assertions for all of these. Same would go for all of the setters that EBean (presumably) uses. This will help from setting invalid values into the model.

Reply:
@bion curious your thoughts given the status quo.

Seemingly the "NOT NULL" database annotations are used to enforce this.

I don't see a problem doing it here also, but wonder if it borders on redundant/errorprone having the restriction in 2 places.

Reply:
While perhaps redundant, I'd put it in the category of "more actionable stack traces", since the error will appear at the time the bad data is provided rather than potentially further down (or even in a separate class) when the actual db operation is attempted.

Reply:
I'd also vote for the "belt and suspenders" approach of enforcing it both in the application code as well as in the DB schema

Reply:
added the setter/ctor checks.  AutoValue already rejects null by default so the serializer doesn't need to worry about that.